rust/kernel/seq_file.rs

```diff
@@ -0,0 +1,225 @@
+//!
+//! This module allows Rust devices to implement [`struct seq_operations`] and
+//! and create a file under `/proc` based on that implementation.
+//!
+//! C header: [`include/linux/seq_file.h`](../../../include/linux/seq_file.h)
+//!
+//! Reference: <https://www.kernel.org/doc/html/latest/filesystems/seq_file.html>
+
+// Currently this module is only usable through proc_fs.
+#![cfg_attr(not(CONFIG_PROC_FS), allow(dead_code))]
+
+use alloc::boxed::Box;
+use core::{
+    iter::{Iterator, Peekable},
+    marker::PhantomData,
+    ptr,
+};
+
+use crate::{bindings, c_types, cstr, CStr};
+
+#[cfg(CONFIG_PROC_FS)]
+pub use proc::proc_create_seq;
+
+/// Rust equivalent of the [`seq_operations`] interface on the C side.
+///
+/// # Example
+///
+/// ```
+/// struct Data(&'static [String]);
+///
+/// impl seq_file::SeqOperations for Data {
+///     type Item = &'static String;
+///     type Iterator = core::slice::Iter<'static, String>;
+///
+///     fn start(arg: &Data) -> Option<Box<Peekable<Self::Iterator>>> {
+///         let iter = arg.0.iter();
+///         Box::try_new(iter.peekable()).ok()
+///     }
+///
+///     fn display(item: &Self::Item) -> &str {
+///         &item[..]
+///     }
+/// }
+/// ```
+///
+/// [`seq_operations`]: ../../../include/linux/seq_file.h
+pub trait SeqOperations {
+    /// Type produced on each iteration.
+    type Item;
+
+    /// Type created when the seq file is opened.
+    type Iterator: Iterator<Item = Self::Item>;
+
+    /// Called once each time the `seq_file` is opened.
+    fn start(arg: &Self) -> Option<Box<Peekable<Self::Iterator>>>;
+
+    /// How the item will be displayed to the reader.
+    fn display(item: &Self::Item) -> &str;
+}
+
+extern "C" fn stop_callback<T: SeqOperations>(
+    _m: *mut bindings::seq_file,
+    v: *mut c_types::c_void,
+) {
+    if !v.is_null() {
+        // SAFETY: `v` was created by a previous call to `next_callback` or
+        // `start_callback` and both functions return either a null pointer
+        // or pointer generated by `Box<Peekable<T::Iterator>>::into_raw`.
+        drop(unsafe { Box::from_raw(v as *mut Peekable<T::Iterator>) })
+    }
+}
+
+extern "C" fn next_callback<T: SeqOperations>(
+    _m: *mut bindings::seq_file,
+    v: *mut c_types::c_void,
+    pos: *mut bindings::loff_t,
+) -> *mut c_types::c_void {
+    if v.is_null() {
+        return ptr::null_mut();
+    }
+
+    // SAFETY: `v` was created by a previous call to `next_callback` or
+    // `start_callback` and both functions return either a null pointer
+    // or pointer generated by `Box<Peekable<T::Iterator>>::into_raw`.
+    // We already checked for he null pointer case above.
+    let mut iterator = unsafe { Box::from_raw(v as *mut Peekable<T::Iterator>) };
+
+    // SAFETY: The caller guarantees tha `pos` is a valid pointer to an
+    // `loff_t` and expects this function to mutate the value.
+    unsafe {
+        *pos += 1;
+    }
+
+    iterator.next();
```

Comment:
Thanks, I didn't realize that iterators could resume like that.

I changed it, but I think in this case we might actually be ok without checking `next`. Since we have a `Peekable` and we've already called `peek` on it and checked for `None` (in an earlier call to `next_callback` or `start_callback`) so `next` will always return `Some` here. See [impl of `next` for `Peekable`](https://doc.rust-lang.org/stable/src/core/iter/adapters/peekable.rs.html#36-41).